### PR TITLE
Collections Docs

### PR DESCRIPTION
**Pull request by josephjclark**
## Short Description

Docs for Collections:

* Under Platform / Build & Manage Workflows we have a high-level Collections overview. This has a couple of code examples but is designed as a user guide, rather than a developer / job writing guide
* Under Adaptors / Collections is a detailed API overview and explanation of how it all works (the important bits anyway)

There is a polite notice indication that collections are in Beta (ie , not totally production ready right now, thanks, but have a play)

## AI Usage

Please disclose how you've used AI in this work (it's cool, we just want to
know!):

- [ ] Code generation (copilot but not intellisense)
- [ ] Learning or fact checking
- [ ] Strategy / design
- [ ] Optimisation / refactoring
- [ ] Translation / spellchecking / doc gen
- [ ] Other
- [x] I have not used AI

You can read more details in our
[Responsible AI Policy](https://www.openfn.org/ai#pull-request-templates)

### CODE REVIEW

File: adaptors/collections.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
+---
+title: Collections Adaptor
+---
+
+## Collections Overview
+
+The Collections API is a key/value storage solution. It is designed for high
```
Comment (aleksa-krolls) [inline adaptors/collections.md:7]: We need to make clear that this is a data store _in_ OpenFn... not something external (like all other adaptors).

File: adaptors/collections.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+performance over a large volume of data.
+
+Use-cases include:
+
+- Storing mapping objects for use in workflows
+- Buffering and aggregating high volumes of incoming data
+- Caching and sharing state between workflows
+
+A Collection is bound to a project. Collections can only be accessed with a
```
Comment (aleksa-krolls) [inline adaptors/collections.md:16]: "a project in OpenFn"?

File: adaptors/collections.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+token associated with that project. When running on the app, a workflow is
```
Comment (aleksa-krolls) [inline adaptors/collections.md:17]: You mean a user's API Key? If yes, we should link to the docs instructions on how to generate one
https://docs.openfn.org/documentation/api-tokens

File: adaptors/collections.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
+---
+title: Collections Adaptor
```
Comment (aleksa-krolls) [inline adaptors/collections.md:2]: Wondering if we can somehow tag `collections` and `common` as special OpenFn adaptors?

File: adaptors/collections.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+under a given key:
+
+You can also pass an array of items for a batch-set.
+
+### Getting data from a collection
+
+To retrieving multiple items from a Collection, we recommend using the `each()`
+function.
+
+`each()` will stream each value individually, greatly reducing the memory
```
Comment (aleksa-krolls) [inline adaptors/collections.md:45]: Do we really have to use `each()`? What about the scenario when I want to query a list of devices that match my search criteria? In some of the use cases we've been imaging so far, we've been getting a long list of records and then sending that entire list to a downstream system, or comparing that list with data from another system. Mtuchi could show you what's going on the openlmis prototype, if interested in learning more.

File: adaptors/collections.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  state.resources ??= {};
+  state.resources[state.data.id] = mappedString;
+  return state;
+});
+```
+
+You can also fetch multiple items with `get()`, which supports the same query
+options as `each()`.
+
+Bear in mind that all the items will be loaded into memory at once. For large
```
Comment (aleksa-krolls) [inline adaptors/collections.md:102]: Hmm... will ping you next week to help me understand tradeoffs of using `get()` and `each()` for different use cases so I can better understand when we might start bumping into limits, and to know which is best for different scenarios... we can then beef up these docs with some examples.

File: adaptors/collections.md (hunk 5 above)
Comment (aleksa-krolls) [reply adaptors/collections.md:45]: I see `get()` now and left a comment below

**Review by aleksa-krolls — CHANGES_REQUESTED** [review]
Major feedback is to make it clear that `collections` is something _in_ openfn and not an external adaptor. 

Then generally, services team can help add more examples and use cases that will help users better imagine how to use this... but we can do that later and debate where those docs go.

**Review by taylordowns2000 — COMMENTED** [review]
Looks good, but I wanted to try out collections on staging just now and was hoping to see how to add items to a collection. Is that documented somewhere else?

**Comment by josephjclark** [comment]
@taylordowns2000 My bad - I forgot to add set values. That's coming up in a minute

File: adaptors/collections.md (hunk 6 above)
Comment (josephjclark) [reply adaptors/collections.md:102]: Don't ping me - let me fix it here!

File: adaptors/collections.md (hunk 3 above)
Comment (josephjclark) [reply adaptors/collections.md:17]: Kinda - thanks for flagging, I'll clean this up

File: adaptors/collections.md (hunk 4 above)
Comment (josephjclark) [reply adaptors/collections.md:2]: Maybe. Collections is definitely special. Common is only sort of special :thinking: 

So I guess to answer this I'd have to ask what we mean by "special" and what exactly we want to flag

**Comment by josephjclark** [comment]
@aleksa-krolls So  to be clear (and the context is probably a bit lost), this is the Overview adaptor docs. If you go to the doc site and the adaptor list, this is the Overview. It's designed as a high level introduction to the collections  feature generally and the collections API specifically.

Collections support is implemented as an adaptor. That's how we inject an API in job code, and how we provide documentation. We should not hide or obfuscate this because it's kind of relevant - the collections adaptor appears in run logs, and if you're using the CLI, you need to pass the collections adaptor explicitly for support.

I agree that we need some more general, less technical docs somewhere in the  main docs page. I'm not sure where. I suppose there's a general `Use Collections` style page which explains the feature at a high level (this can show API examples and will link to the adaptor docs, but otherwise doesn't treat collections as an adaptor), plus a `Collections Admin` page which explains how to create, delete and a generally manage collections through the admin API.

But I don't know WHERE that goes. Maybe under Platform as a section called Collections? Or nested under Platform / Build & Manage Workflows (that seems very deeply nested, but I think it makes sense as a peer to eg Credentials, Limits, Snaphots, Triggers)

**Comment by aleksa-krolls** [comment]
@josephjclark I agree it makes sense there. Can we do that to start, and can you please raise in engineering standup to get this done sometime in the coming days? This page should then link to the adaptor docs. 
> But I don't know WHERE that goes. Maybe under Platform as a section called Collections? Or nested under Platform / Build & Manage Workflows (that seems very deeply nested, but I think it makes sense as a peer to eg Credentials, Limits, Snaphots, Triggers)

Screenshot 2024-11-07 at 9 05 57 PM

**Comment by josephjclark** [comment]
@aleksa-krolls I'm taking responsibility for docs on this, so I'll get it done tomorrow.

**Comment by josephjclark** [comment]
So:

* The Adaptors page is now a technical overview and API reference for the Collections API.
* The Docs page is a high level overview of the feature. It has a couple of code examples and refers to the API, but otherwise is a conceptual overview and usage guide more than a technical one.

@stuartc Would you mind looking over both pages and calling out any egregious errors?

These docs DO NOT include:

* performance tips for big data sets (eg, use each not get). I would like to do this later but I just don't have time and I'm cutting cloth
* Details on usage limits and costs. I have no idea what the rules are here
* A good sample workflow which uses collections. Not that we have a good place to share this at the moment...

Please also note that I've raised an issue for better collections support in the CLI (eg, `openfn collections set` would be awesome), se https://github.com/OpenFn/kit/issues/819

File: docs/build/collections.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+sidebar_label: Collections
+---
+
+Collections provides a high-volume, high-performance storage solution built into
+OpenFn.
+
+Collections is suitable for buffering, caching and aggregating data from
+Webhooks, storing large mapping files, and sharing state between workflows.
+
+Collections can be used to store a very large number of items (in the order of
```
Comment (josephjclark) [inline docs/build/collections.md:12]: @stuartc your nervousness please about the phrase `in the order of millions`

I just want give some idea of what "high volume" means. I know it's woolly.

Comment (stuartc) [reply docs/build/collections.md:12]: I think we can leave it as is for now. We will need to determine two things:

1. How much work it is for the db to store really large values, right now they are strings and not indexed, but can be really large; save for our middleware that will reject large (>10MB) HTTP requests.
2. If users are ok without being able to query by value.

If either of these are an issue, then the number of records we would want to limit would be clearer. But for now, millions should be ok.

**Review by stuartc — APPROVED** [review]
Wonderful! 🫶

**Review by aleksa-krolls — APPROVED** [review]

**Pull request merged**